File: 02_02_2024.py
class Solution:
    def atoi(self, string):
        if not string:
            return -1

        n = len(string)
        result = 0
        sign = 1

        # Skip leading whitespaces
        i = 0
        while i < n and string[i] == ' ':
            i += 1

        # Check for optional sign
        if i < n and (string[i] == '-' or string[i] == '+'):
            sign = -1 if string[i] == '-' else 1
            i += 1

        # Process digits
        while i < n and string[i].isdigit():
            digit = int(string[i])
            result = result * 10 + digit
            i += 1

        # Apply sign to the result
        result *= sign

        # Check for overflow
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        if result > INT_MAX:
            return INT_MAX
        elif result < INT_MIN:
            return INT_MIN

        # Check for remaining non-digit characters
        while i < n and not string[i].isdigit():
            return -1

        return result
        

File: 02_02_2024.cpp
//User function template for C++

class Solution {
public:
    int atoi(string s) {
        int n = s.length();
        if (n == 0) return -1;

        int i = 0;
        int sign = 1;
        int result = 0;

        // Handling whitespaces
        while (i < n && s[i] == ' ') {
            i++;
        }

        // Handling sign
        if (i < n && (s[i] == '+' || s[i] == '-')) {
            sign = (s[i++] == '-') ? -1 : 1;
        }

        // Constructing the number
        while (i < n && isdigit(s[i])) {
            // Handling overflow
            if (result > INT_MAX / 10 || (result == INT_MAX / 10 && s[i] - '0' > 7)) {
                return (sign == 1) ? INT_MAX : INT_MIN;
            }

            result = result * 10 + (s[i++] - '0');
        }

        // Checking for non-numeric characters
        while (i < n && !isdigit(s[i])) {
            return -1;
        }

        return result * sign;
    }
};

//
